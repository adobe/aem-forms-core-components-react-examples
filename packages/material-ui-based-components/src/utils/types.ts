import { FieldJson, FieldsetJson, State } from '@aemforms/af-core';
import { Handlers, WithViewState } from '@aemforms/af-react-renderer';

export type FieldViewState = WithViewState<FieldJson>;

export type PROPS = State<FieldJson & Handlers & {
  isError?: boolean,
  layout?: {
    [key: string]: any
  }
  text?:string,
  richText?:boolean
}>;

export type PROPS_PANEL = State<FieldsetJson> & { handlers: Handlers }
