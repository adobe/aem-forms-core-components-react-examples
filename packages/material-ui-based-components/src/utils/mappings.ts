import Form from '../components/Form';
import Panel from '../components/Panel';
import TextField from '../components/TextField';
import Date from '../components/Date';
import PlainText from '../components/PlainText';
import Title from '../components/Title';
import Image from '../components/Image';
import Rating from '../components/Rating';
import Slider from '../components/Slider';
import TabsVerticalComp from '../components/TabsVerticalComp';
import TabsHorizontalComp from '../components/TabsHorizontalComp';

export default {
  form: Form,
  panel: Panel,
  'text-input': TextField,
  'date-input': Date,
  'text-input': TextField,
  'plain-text': PlainText,
  'custom:title': Title,
  image: Image,
  'text-input': TextField,
  'custom:rating':Rating,
  'custom:slider':Slider
  'text-input': TextField,
  'custom:vertical-tab': TabsVerticalComp,
  'custom:horizontal-tab': TabsHorizontalComp
};