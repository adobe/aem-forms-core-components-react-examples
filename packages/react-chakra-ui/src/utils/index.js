import React from 'react';
import { FormContext } from '@aemforms/af-react-renderer';
import { createFormInstance } from '@aemforms/af-core';
import { ChakraProvider } from '@chakra-ui/react'
import { IntlProvider } from 'react-intl';
import { render } from '@testing-library/react';

export const DEFAULT_ERROR_MESSAGE = 'There is an error in the field';

export const createForm = (field) => {
  const formJson = {
    items: [field]
  };
  return createFormInstance(formJson);
};

export const Provider = (form, mappings, locale, dictionaries) => props => {
  const c = {
    form,
    mappings,
    modelId: '$form',
    refMap: {}
  };
  const { children } = props;
  return (
    <ChakraProvider>
      <IntlProvider locale={locale} messages={dictionaries}>
        <FormContext.Provider value={c}>{children}</FormContext.Provider>
      </IntlProvider>
    </ChakraProvider>
  );
};

export const renderComponent = function (Component) {
  const test = (field, operation) => {
    const form = createForm(field);
    if (operation) {
      operation(form, form.items[0]);
    }
    const e = form.items[0].getState();
    let component = <Component {...e} />;
    const wrapper = Provider(form);
    const renderResponse = render(component, { wrapper });
    return {
      renderResponse,
      form,
      element: form?.items[0],
      component
    };
  };
  return test;
};
