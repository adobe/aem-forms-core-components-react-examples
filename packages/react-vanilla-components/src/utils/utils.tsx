export const formatBytes = (bytes: number, decimals = 0) => {
    if (!+bytes) {
      return '0 Bytes';
    }
    const k = 1024;
    const dm = decimals < 0 ? 0 : decimals;
    const sizes = ['bytes', 'kb', 'mb', 'gb', 'tb', 'pb'];
    const i = Math.floor(Math.log(bytes) / Math.log(k));

    return `${parseFloat((bytes / Math.pow(k, i)).toFixed(dm))} ${sizes[i]}`;
  };

  export const syncAriaDescribedBy = (id: string, tooltip?: string, description?: string, error?: string) => {
    const descriptions = [];

    if (description) {
        descriptions.push(`${id}__longdescription`);
    }
    if (tooltip) {
        descriptions.push(`${id}__shortdescription`);
    }
    if (error) {
        descriptions.push(`${id}__errormessage`);
    }

    return descriptions.join(' ');
};