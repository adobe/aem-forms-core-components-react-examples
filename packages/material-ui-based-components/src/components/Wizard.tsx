import React, { useCallback, useContext } from 'react';
import Box from '@mui/material/Box';
import Stepper from '@mui/material/Stepper';
import Step from '@mui/material/Step';
import StepLabel from '@mui/material/StepLabel';
import Button from '@mui/material/Button';
import { withRuleEnginePanel } from '../shared/withRuleEngine';
import { PROPS_PANEL, TabPanelProps } from '../utils/types';
import { FormContext, getRenderer } from '@aemforms/af-react-renderer';

function TabPanel(props: TabPanelProps) {
    const { children, value, index, ...other } = props;

    return (
        <div
            hidden={value !== index}
            id={children.id}
            {...other}
        >
            {value === index && (
                <Box sx={{ p: 3 }}>
                    {children}
                </Box>
            )}
        </div>
    );
}

function Wizard(props: PROPS_PANEL) {
    //@ts-ignore
    const mappings = useContext(FormContext).mappings;
    const items = props.visible ? props.items : [];
    const [activeStep, setActiveStep] = React.useState(0);

    const steps = items.map((child: any) => (
        child.label.value
    ));

    const getChild = useCallback((child: any, index: number) => {
        const Comp = getRenderer(child, mappings);
        return Comp ? <Comp key={`${child.id}_${index}`} {...child} /> : (null);
    }, [mappings]);

    const handleNext = useCallback(() => {
        setActiveStep((prevActiveStep) => prevActiveStep + 1);
        window.scrollTo(0, 0);
    }, []);

    const handleBack = useCallback(() => {
        setActiveStep((prevActiveStep) => prevActiveStep - 1);
        window.scrollTo(0, 0);
    }, []);

    return (
        <Box sx={{ width: '100%' }}>
            <Stepper activeStep={activeStep}>
                {steps.map((label) => {
                    const stepProps: { completed?: boolean } = {};
                    return (
                        <Step key={label} {...stepProps}>
                            {/* @ts-ignore */}
                            <StepLabel>{label}</StepLabel>
                        </Step>
                    );
                })}
            </Stepper>

            <React.Fragment>
                {
                    items.map((child: any, index: number) => (
                        <TabPanel value={activeStep} index={index} key={`${child.id}`}>
                            {getChild(child, index)}
                        </TabPanel>
                    ))
                }
                <Box sx={{ display: 'flex', flexDirection: 'row', pt: 2 }}>
                    <Button
                        color="inherit"
                        disabled={activeStep === 0}
                        onClick={handleBack}
                    >
                        Back
                    </Button>
                    <Box sx={{ flex: '1 1 auto' }} />
                    <Button
                        color="inherit"
                        disabled={activeStep === steps.length - 1}
                        onClick={handleNext}
                    >
                        Next
                    </Button>
                </Box>
            </React.Fragment>

        </Box>
    );
}

export default withRuleEnginePanel(Wizard);