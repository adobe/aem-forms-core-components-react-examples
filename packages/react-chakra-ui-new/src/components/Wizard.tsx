/*************************************************************************
* ADOBE CONFIDENTIAL
* ___________________
*
* Copyright 2023 Adobe
* All Rights Reserved.
*
* NOTICE: All information contained herein is, and remains
* the property of Adobe and its suppliers, if any. The intellectual
* and technical concepts contained herein are proprietary to Adobe
* and its suppliers and are protected by all applicable intellectual
* property laws, including trade secret and copyright laws.
* Dissemination of this information or reproduction of this material
* is strictly forbidden unless prior written permission is obtained
* from Adobe.

* Adobe permits you to use and modify this file solely in accordance with
* the terms of the Adobe license agreement accompanying it.
*************************************************************************/

import React, { useCallback, useContext } from 'react';
import { Button, Box, Tabs, TabList, TabPanels, Tab, TabPanel } from '@chakra-ui/react';
import { withRuleEnginePanel } from '../shared/withRuleEngine';
import { PROPS_PANEL } from '../utils/type';
import { FormContext, getRenderer } from '@aemforms/af-react-renderer';

function Wizard(props: PROPS_PANEL) {
    const mappings = useContext(FormContext).mappings;
    const items = props.visible ? props.items : [];
    const [activeStep, setActiveStep] = React.useState(0);

    const steps = items.map((child: any) => (
        child.label.value
    ));

    const getChild = useCallback((child: any, index: number) => {
        const Comp = getRenderer(child, mappings);
        return Comp ? <Comp key={`${child.id}_${index}`} {...child} /> : (null);
    }, [mappings]);

    const handleNext = useCallback(() => {
        setActiveStep((prevActiveStep) => prevActiveStep + 1);
    }, []);

    const handleBack = useCallback(() => {
        setActiveStep((prevActiveStep) => prevActiveStep - 1);
    }, []);

    return (
            <Tabs isFitted index={activeStep}>
                <TabList>
                    {steps.map((label) => {
                        const stepProps: { completed?: boolean } = {};
                        return (
                            <Tab sx={{ opacity: '1 !important', cursor: 'inherit !important' }} isDisabled key={label} {...stepProps}>
                                {label}
                            </Tab>
                        );
                    })}
                </TabList>
                <Box sx={{ display: 'flex', flexDirection: 'row', margin: '5px' }}>
                    <Button
                        isDisabled={activeStep === 0}
                        onClick={handleBack}
                    >
                        Back
                    </Button>
                    <Box sx={{ flex: '1 1 auto' }} />
                    <Button
                        isDisabled={activeStep === steps.length - 1}
                        onClick={handleNext}
                    >
                        Next
                    </Button>
                </Box>
                <TabPanels>
                {
                    items.map((child: any, index: number) => (
                        <TabPanel index={index} key={`${child.id}`}>
                            {getChild(child, index)}
                        </TabPanel>
                    ))
                }
                </TabPanels>
            </Tabs>
    );
}

export default withRuleEnginePanel(Wizard);