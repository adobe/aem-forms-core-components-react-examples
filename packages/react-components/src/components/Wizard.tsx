// *******************************************************************************
//  * Copyright 2022 Adobe
//  *
//  * Licensed under the Apache License, Version 2.0 (the “License”);
//  * you may not use this file except in compliance with the License.
//  * You may obtain a copy of the License at
//  *
//  *     http://www.apache.org/licenses/LICENSE-2.0
//  *
//  * Unless required by applicable law or agreed to in writing, software
//  * distributed under the License is distributed on an “AS IS” BASIS,
//  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  * See the License for the specific language governing permissions and
//  * limitations under the License.

//  * The BEM markup is as per the AEM core form components guidelines.
//  * LINK- https://github.com/adobe/aem-core-forms-components/blob/master/ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/wizard/v1/wizard/wizard.html
//  ******************************************************************************

import React, { useState, useContext, useCallback } from 'react';
import {FormContext} from '@aemforms/af-react-renderer';
import {getRenderer} from '@aemforms/af-react-renderer';
import { withRuleEnginePanel }  from '../utils/withRuleEngine';
import { PROPS_PANEL } from '../utils/type';

const Wizard = (props: PROPS_PANEL) => {
   const {mappings} = useContext(FormContext);
   const { items, label, id, visible, enabled } = props;
   const [activeButtonIndex, setActiveButtonIndex] = useState(0);
   const [shortDescription, setShortDescription] = useState(true);
   const [longDescription, setLongtDescription] = useState(false);

   const {v: visibleItems} =
   items.reduce(({v} : any, item) => {
       const isVisible = item.visible === true;
       return {
           v: isVisible ? v.concat([item]) : v
       };
   }, {v: []});

 const handlePrevious = () => {
    const newIndex = activeButtonIndex - 1;
    newIndex >= 0 && setActiveButtonIndex(newIndex < 0 ? visibleItems.length - 1 : newIndex);
 };

 const handleNext = () => {
    const newIndex = activeButtonIndex + 1;
    newIndex <= visibleItems.length - 1 && setActiveButtonIndex(newIndex >= visibleItems.length ? 0 : newIndex);
 };

 const handleClick = (event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
   event.preventDefault();  
   setShortDescription(!shortDescription);
   setLongtDescription(!longDescription);
};

const getChild = useCallback((child: any, index: number) => {
   const Comp = getRenderer(child, mappings);
   return Comp ? <Comp key={`${child.id}_${index}`} {...child} /> : (null);
}, []);

//  const Comp = visibleItems.length > 0 ? getRenderer(visibleItems[activeButtonIndex], mappings) : '';

 return visible ? (
    <div id={id} className="cmp-adaptiveform-wizard" data-panelcontainer="wizard" data-placeholder-text="Please drag Wizard components here" data-cmp-visible={visible} data-cmp-enabled={enabled}>
         {label?.visible && <label className="cmp-adaptiveform-wizard__label" htmlFor={`${id}-widget`}>{label?.value}</label>}
         {props?.description && <button className="cmp-adaptiveform-wizard__questionmark" onClick={handleClick} aria-label='Toggle Button'></button>} 
         {shortDescription && props?.tooltip && <div title='Help Text' data-cmp-visible={shortDescription} className='cmp-adaptiveform-wizard__shortdescription'>{props?.tooltip}</div>}
          <div aria-live="polite">
             {longDescription && props?.description && <div title='Help Text' data-cmp-visible={longDescription} className="cmp-adaptiveform-wizard__longdescription">{props?.description}</div>}
         </div>
           <div className="cmp-adaptiveform-wizard__widget cmp-adaptiveform-wizard__widget--empty" id={`${id}-widget`}>
            <div className="cmp-adaptiveform-wizard__tabs-container">
                <ol id={`${id}_guide-item-nav-container`} className="cmp-adaptiveform-wizard__tabList" role="tablist">
                   {visibleItems.map((item: PROPS_PANEL, index: number) =>( 
                   <li key={item.id} title={item?.label?.value} id={`${item.id}_wizard-item-nav`} className={index===activeButtonIndex?'cmp-adaptiveform-wizard__tab cmp-adaptiveform-wizard__tab--active': 'cmp-adaptiveform-wizard__tab' } role="tab" tabIndex={index===activeButtonIndex?0:-1} aria-selected={index===activeButtonIndex?'true':'false'} data-cmp-hook-adaptiveformwizard="tab">{item.label?.value}</li>
                    ))}
               </ol>
           </div>
           {visibleItems.length > 0 ? <div className="cmp-adaptiveform-wizard__previousNav" role="navigation" aria-label="Previous Button" onClick={handlePrevious} tabIndex={0}></div> : null}
           {visibleItems.length > 0 ? <div data-cmp-hook-adaptiveformtabs="tabpanel" className={activeButtonIndex === visibleItems[activeButtonIndex].index?'cmp-adaptiveform-wizard__wizardpanel cmp-adaptiveform-wizard__wizardpanel--active':'cmp-adaptiveform-wizard__wizardpanel'} role="tabpanel" tabIndex={0} aria-hidden={activeButtonIndex===visibleItems[activeButtonIndex].index?'false':'true'} aria-labelledby={`${visibleItems[activeButtonIndex].id}_wizard-item-nav`} data-cmp-hook-adaptiveformwizard="wizardpanel">
              {/* <Comp key={`${visibleItems[activeButtonIndex].id}_${activeButtonIndex}`} {...visibleItems[activeButtonIndex]} /> */}
              {getChild(visibleItems[activeButtonIndex], activeButtonIndex)}
           </div> : null}
           {visibleItems.length > 0 ? <div className="cmp-adaptiveform-wizard__nextNav" role="navigation" aria-label="Next Button" onClick={handleNext} tabIndex={0}> </div> : null}
          </div>    
    </div>
  ): null;
};

export default withRuleEnginePanel(Wizard);