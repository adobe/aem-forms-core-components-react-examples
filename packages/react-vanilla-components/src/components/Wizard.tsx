// *******************************************************************************
//  * Copyright 2023 Adobe
//  *
//  * Licensed under the Apache License, Version 2.0 (the “License”);
//  * you may not use this file except in compliance with the License.
//  * You may obtain a copy of the License at
//  *
//  *     http://www.apache.org/licenses/LICENSE-2.0
//  *
//  * Unless required by applicable law or agreed to in writing, software
//  * distributed under the License is distributed on an “AS IS” BASIS,
//  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  * See the License for the specific language governing permissions and
//  * limitations under the License.

//  * The BEM markup is as per the AEM core form components guidelines.
//  * LINK- https://github.com/adobe/aem-core-forms-components/blob/master/ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/wizard/v1/wizard/wizard.html
//  ******************************************************************************

import React, { useState, useContext, useCallback } from 'react';
import { FormContext } from '@aemforms/af-react-renderer';
import { getRenderer } from '@aemforms/af-react-renderer';
import { withRuleEnginePanel } from '../utils/withRuleEngine';
import { PROPS_PANEL } from '../utils/type';
import FieldWrapper from './common/FieldWrapper';

const Wizard = (props: PROPS_PANEL) => {
  // @ts-ignore
  const { mappings } = useContext(FormContext);
  const { items, label, id, visible, enabled, appliedCssClassNames } = props;
  const [activeTab, setActiveTab] = useState(0);

  const { v: visibleItems } =
    items.reduce(({ v }: any, item) => {
      const isVisible = item.visible === true;
      return {
        v: isVisible ? v.concat([item]) : v
      };
    }, { v: [] });


  const previousHandler = useCallback(() => {
    setActiveTab(activeTab - 1);
  }, [activeTab]);

  const nextHandler = useCallback(() => {
    setActiveTab(activeTab + 1);
  }, [activeTab]);

  const getChild = useCallback((child: any, index: number) => {
    const Comp = getRenderer(child, mappings);
    return Comp ? <Comp key={`${child.id}_${index}`} {...child} /> : (null);
  }, []);

  return visible ? (
    <div
      id={id}
      className={`cmp-adaptiveform-wizard ${appliedCssClassNames || ''}`}
      data-panelcontainer="wizard"
      data-placeholder-text="Please drag Wizard components here"
      data-cmp-visible={visible}
      data-cmp-enabled={enabled}
    >
      <FieldWrapper
        bemBlock='cmp-adaptiveform-wizard'
        label={label}
        id={id}
        tooltip={props.tooltip}
        description={props.description}
      />
      <div
        className="cmp-adaptiveform-wizard__widget cmp-adaptiveform-wizard__widget--empty"
        id={`${id}-widget`}
      >
        <div className="cmp-adaptiveform-wizard__tabs-container">
          <ol id={`${id}_guide-item-nav-container`} className="cmp-adaptiveform-wizard__tabList" role="tablist">
            {visibleItems.map((item: PROPS_PANEL, index: number) => (
              <li
                key={item.id}
                title={item?.label?.value}
                id={`${item.id}_wizard-item-nav`}
                className={`cmp-adaptiveform-wizard__tab ${index === activeTab ? 'cmp-adaptiveform-wizard__tab--active' : ''}`}
                role="tab"
                tabIndex={index === activeTab ? 0 : -1}
                aria-selected={index === activeTab ? 'true' : 'false'}
                data-cmp-hook-adaptiveformwizard="tab"
              >
                {item.label?.value}
              </li>
            ))}
          </ol>
        </div>
        <div className="cmp-adaptiveform-wizard__containerNav">
          {
            visibleItems.length > 0 && activeTab > 0 ?
              <div
                className="cmp-adaptiveform-wizard__nav cmp-adaptiveform-wizard__nav--previous"
                role="navigation"
                aria-label="Previous Button"
                onClick={previousHandler}
                tabIndex={0}
              />
              : null
          }
          {
            visibleItems.length > 0 && activeTab < visibleItems.length - 1 ?
              <div
                className="cmp-adaptiveform-wizard__nav cmp-adaptiveform-wizard__nav--next"
                role="navigation"
                aria-label="Next Button"
                onClick={nextHandler}
                tabIndex={0}
              />
              : null
          }
        </div>
        {
          visibleItems.length > 0 ?
            <div
              data-cmp-hook-adaptiveformtabs="tabpanel"
              className={`cmp-adaptiveform-wizard__wizardpanel ${activeTab === visibleItems[activeTab].index ? 'cmp-adaptiveform-wizard__wizardpanel--active' : ''}`}
              role="tabpanel"
              tabIndex={0}
              aria-hidden={activeTab === visibleItems[activeTab].index ? 'false' : 'true'}
              aria-labelledby={`${visibleItems[activeTab].id}_wizard-item-nav`}
              data-cmp-hook-adaptiveformwizard="wizardpanel"
            >
              {getChild(visibleItems[activeTab], activeTab)}
            </div>
            : null
        }
      </div>
    </div>
  ) : null;
};

export default withRuleEnginePanel(Wizard);