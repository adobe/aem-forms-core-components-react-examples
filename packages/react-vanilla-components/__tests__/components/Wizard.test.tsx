/*
 * Copyright 2023 Adobe, Inc.
 *
 * Your access and use of this software is governed by the Adobe Customer Feedback Program Terms and Conditions or other Beta License Agreement signed by your employer and Adobe, Inc.. This software is NOT open source and may not be used without one of the foregoing licenses. Even with a foregoing license, your access and use of this file is limited to the earlier of (a) 180 days, (b) general availability of the product(s) which utilize this software (i.e. AEM Forms), (c) January 1, 2023, (d) Adobe providing notice to you that you may no longer use the software or that your beta trial has otherwise ended.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL ADOBE NOR ITS THIRD PARTY PROVIDERS AND PARTNERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import Wizard from "../../src/components/Wizard";
import React from "react";
import { render } from "@testing-library/react";
import { renderComponent } from "../utils";
import userEvent from "@testing-library/user-event";

const WizardWithData = {
  id: "wizard",
  ":type": "core/fd/components/form/wizard/v1/wizard",
  visible: true,
  label: { value: "Wizard" },
  index: 0,
  items: [
    {
      id: "field",
      fieldType: "text-field",
      visible: true,
      index: 0,
      label: { value: "text field label" },
    },
    {
      id: "tabsWithData",
      ":type": "core/fd/components/form/tabsontop/v1/tabsontop",
      visible: true,
      label: { value: "tab label" },
      index: 0,
      items: [
        {
          id: "field",
          name: "f1",
          fieldType: "text-field",
          visible: true,
          index: 0,
          label: { value: "text field label" },
        },
        {
          id: "field2",
          name: "f2",
          fieldType: "text-field",
          visible: true,
          index: 1,
          label: { value: "text field 2 label" },
        },
      ],
    }
  ],
};

const emptyWizard = {
  id: "emptyWizard",
  ":type": "core/fd/components/form/wizard/v1/wizard",
  visible: true,
  label: { value: "empty label" },
  items: [],
  index: 0,
};

test("clicking on the next tab should switch from one component to another and on clicking the previous button it should take to the previous component",() => {
  const helper = renderComponent(Wizard);
  const { renderResponse } = helper(WizardWithData);
  const NextTab = renderResponse.container.getElementsByClassName(
    "cmp-adaptiveform-wizard__nextNav"
  );
  const PrevTab = renderResponse.container.getElementsByClassName(
    "cmp-adaptiveform-wizard__previousNav"
  );
  expect(NextTab.length).toEqual(1);
  expect(PrevTab.length).toEqual(0);
  const tablist = renderResponse.getAllByRole('tab');
  expect(tablist).toHaveLength(2);
  expect(tablist[0].className.includes('cmp-adaptiveform-wizard__tab cmp-adaptiveform-wizard__tab--active'))
  expect(tablist[1].className.includes('cmp-adaptiveform-wizard__tab'))
  userEvent.click(NextTab[0])
  expect(NextTab.length).toEqual(0)
  expect(tablist[0].className.includes('cmp-adaptiveform-wizard__tab'))
  expect(tablist[1].className.includes('cmp-adaptiveform-wizard__tab cmp-adaptiveform-wizard__tab--active'))
  expect(PrevTab.length).toEqual(1);
});

test("Tabs should not rendered if all the items are not visible", () => {
  const helper = renderComponent(Wizard);
  const sample = {
    ...WizardWithData,
    items: WizardWithData.items.map((x) => ({
      ...x,
      visible: false,
    })),
  };
  const { renderResponse } = helper(sample);
  expect(renderResponse.queryByText(sample.label.value)).not.toBeNull();
  const tabs = renderResponse.queryAllByRole("navigation");
  expect(tabs.length).toEqual(0);
});

test("Hidden component should not be rendered",() => {
  const helper = renderComponent(Wizard);
  const sample = {
    ...WizardWithData,
    items: [
      {
        ...WizardWithData.items[0],
        visible: false,
      }
    ],
  };
  const { renderResponse } = helper(sample);
  const tabs = renderResponse.container.querySelectorAll('[role="navigation"]');
  expect(tabs.length).toEqual(0);
  let tabPanels = renderResponse.container.querySelectorAll(".f1");
  expect(tabPanels.length).toEqual(0);
});

test("Tabs should not rendered if item length is zero", () => {
  const helper = renderComponent(Wizard);
  const { renderResponse } = helper(emptyWizard);
  expect(renderResponse.queryByText(emptyWizard.label.value)).not.toBeNull();
  const tabs = renderResponse.queryAllByRole("navigation");
  expect(tabs.length).toEqual(0);
  const tabPanels = renderResponse.queryAllByRole("tabpanel");
  expect(tabPanels.length).toEqual(0);
});

test("Tabs should not rendered if visible is false", () => {
  const sample = {
    ...emptyWizard,
    visible: false,
  };
  const { container } = render(<Wizard {...sample} />);
  expect(container.innerHTML.length).toEqual(0);
});

test('In case of both tooltip and description, tooltip should be visible and onclick of toggle button, description should be visible', () => {
  const f = {
    ...WizardWithData,
    tooltip: 'Short Description',
    description: 'Mandatory'
  };
  const helper = renderComponent(Wizard);
  const { renderResponse } = helper(f);
  expect(renderResponse.getByText('Short Description')).not.toBeNull();
  const button = renderResponse.container.getElementsByClassName('cmp-adaptiveform-wizard__questionmark');
  userEvent.click(button[0]);
  expect(renderResponse.getByText('Mandatory')).not.toBeNull();
});
