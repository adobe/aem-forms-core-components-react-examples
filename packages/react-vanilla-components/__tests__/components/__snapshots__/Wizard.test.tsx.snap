// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
<div>
  <div
    class="cmp-adaptiveform-wizard cmp-adaptiveform-wizard--v2 "
    data-cmp-enabled="true"
    data-cmp-visible="true"
    data-panelcontainer="wizard"
    data-placeholder-text="Please drag Wizard components here"
    id="wizard"
  >
    <div
      class="cmp-adaptiveform-wizard__label-container"
    >
      <label
        class="cmp-adaptiveform-wizard__label"
        for="wizard-widget"
      >
        Wizard
      </label>
    </div>
    <div
      aria-live="polite"
    />
    <div
      class="cmp-adaptiveform-wizard__widget cmp-adaptiveform-wizard__widget--v2"
      id="wizard-widget"
    >
      <div
        class="cmp-adaptiveform-wizard__tabs-container"
      >
        <ol
          class="cmp-adaptiveform-wizard__tabList"
          id="wizard_guide-item-nav-container"
          role="tablist"
        >
          <li
            aria-selected="true"
            class="cmp-adaptiveform-wizard__tab cmp-adaptiveform-wizard__tab--active"
            data-cmp-hook-adaptiveformwizard="tab"
            id="field_wizard-item-nav"
            role="tab"
            tabindex="0"
            title="text field label"
          >
            text field label
          </li>
          <li
            aria-selected="false"
            class="cmp-adaptiveform-wizard__tab "
            data-cmp-hook-adaptiveformwizard="tab"
            id="tabsWithData_wizard-item-nav"
            role="tab"
            tabindex="-1"
            title="tab label"
          >
            tab label
          </li>
        </ol>
      </div>
      <div
        class="cmp-adaptiveform-wizard__containerNav"
      >
        <div
          aria-label="Next Button"
          class="cmp-adaptiveform-wizard__nav cmp-adaptiveform-wizard__nav--next"
          role="navigation"
          tabindex="0"
        />
      </div>
      <div
        aria-hidden="false"
        aria-labelledby="field_wizard-item-nav"
        class="cmp-adaptiveform-wizard__wizardpanel cmp-adaptiveform-wizard__wizardpanel--v2 cmp-adaptiveform-wizard__wizardpanel--active"
        data-cmp-hook-adaptiveformtabs="tabpanel"
        data-cmp-hook-adaptiveformwizard="wizardpanel"
        role="tabpanel"
        tabindex="0"
      />
    </div>
  </div>
</div>
`;
